conn.go

```diff
@@ -78,7 +78,9 @@ func createConn(nextConn net.Conn, flightHandler flightHandler, handshakeMessage
 		return nil, errNoConfigProvided
 	case nextConn == nil:
 		return nil, errNilNextConn
-	case config.Certificate != nil && (config.PSK != nil || config.PSKIdentityHint != nil):
+	case (config.PSK == nil) != (config.PSKIdentityHint == nil):
```

Comment:
¿PSKIdentityHint is always mandatory?

>  To help the client in selecting which identity to use, the server can provide a "PSK identity hint" in the ServerKeyExchange message.  If no hint is provided, the ServerKeyExchange message is omitted.

https://tools.ietf.org/html/rfc4279#section-2

Reply:
oh nice catch, do you know if `ClientKeyExchange` is mandatory?

Reply:
Did a quick search for `ClientKeyExchange` through the RFC, no mention of it being optional in certain cases. Though mind you, there is also this:

>  When these ciphersuites are used, the ServerKeyExchange and
>  ClientKeyExchange messages also include the Diffie-Hellman
>  parameters.  The PSK identity and identity hint fields have the same
>  meaning as in the previous section (note that the ServerKeyExchange
>  message is always sent, even if no PSK identity hint is provided).
https://tools.ietf.org/html/rfc4279#section-3

That's only for DHE_PSK though, which is not what we're doing here.

Reply:
@igolaizola @daenney fixed https://github.com/pion/dtls/pull/76/commits/42310ae98b4c0c9a71753bd381a9421df11cde1a!

`DHE_PSK` probably will not be easy to add :/ currently our Handshake parsing is stateless which makes it a little harder to parse things. Right now it is easy to distinguish DHE vs PSK, but the more we add we will have to think about refactoring.

Personally I would love to see someone majorly refactor stuff. It would be great if someone went in with fresh eyes and moved stuff around. I would like to see a bunch of small packages in `internal` instead of the bloated main directory (my 2 cents)

Reply:
`ClientKeyExchange` is mandatory and `ServerKeyExchange` optional depending if a identity hint has been provided (assuming DHE is not supported).

Reply:
`DHE` isn't supported (yet)

I added a test to assert ^ https://github.com/pion/dtls/pull/76/files#diff-f9df1a974544ddcc7ccada9d10f705a4R185

It does a full test with ClientIdentity nil and non-nil (and we assert the values match up)